README.md

<!-- README.md is generated from README.Rmd. Please edit that file -->

# shinypages

<!-- badges: start -->

[![Lifecycle:
maturing](https://img.shields.io/badge/lifecycle-maturing-blue.svg)](https://lifecycle.r-lib.org/articles/stages.html#maturing)
[![R-CMD-check](https://github.com/gorcha/shinypages/actions/workflows/R-CMD-check.yaml/badge.svg)](https://github.com/gorcha/shinypages/actions/workflows/R-CMD-check.yaml)
[![Codecov test
coverage](https://codecov.io/gh/gorcha/shinypages/branch/main/graph/badge.svg)](https://app.codecov.io/gh/gorcha/shinypages?branch=main)
<!-- badges: end -->

shinypages is a simple multi-page router for Shiny applications, very
much inspired by Appsilon’s
[shiny.router](https://github.com/Appsilon/shiny.router).

## Installation

You can install the development version of shinypages like so:

``` r
# install.packages("devtools")
devtools::install_github("gorcha/shinypages")
```

## Example

``` r
library(shiny)
library(shinypages)

# Create a simple page with only UI
ui_home <- function(id) {
  tagList(
    h2("Hello world!"),
    p(a(href = router_link("cars"), "Cars")),
    p(a(href = router_link("admin"), "Admin"))
  )
}

home_page <- router_page("home", ui_home)


# Create a page with UI and a module server
ui_cars <- function(id) {
  ns <- NS(id)
  tagList(
    h2("Cars"),
    p(a(href = router_link("home"), "Home")),
    tableOutput(ns("cars"))
  )
}

server_cars <- function(id) {
  moduleServer(id, function(input, output, session) {
    output$cars <- renderTable(mtcars)
  })
}

cars_page <- router_page("cars", ui_cars, server_cars)

# Restrict access to a page
is_admin <- function(session) {
  "admin" %in% session$groups
}

ui_admin <- function(id) {
  tagList(
    h2("Top Secret"),
    p(a(href = router_link("home"), "Home"))
  )
}

admin_page <- router_page("admin", ui_admin, authorised = is_admin)

# Create a router
my_router <- router(home_page, cars_page, admin_page)

# Add your router into a Shiny application
ui <- fixedPage(
  router_ui()
)

server <- function(input, output, session) {
  router_server(my_router, input, output, session)
}

shinyApp(ui, server)
```


NEWS.md
# shinypages (development version)

* Improved documentation.

* `router()` and `router_page()` objects gain `print()` methods.

* The `title` argument to `router_page()` now defaults to `NULL` instead of
  `""`.
  
* New `get_router_page_metadata()` retrieves the metadata list for the current
  page.

# shinypages 0.1.0

* Initial release.


tests/testthat.R
library(testthat)
library(shinypages)

test_check("shinypages")


tests/testthat/test-shiny.R

test_that("router_ui", {
  expect_s3_class(router_ui(), "shiny.tag")
})

test_that("router_server", {
  session <- list(
    clientData = list(url_hash = "home"),
    userData = list()
  )

  rt <- router(router_page("", function(id) htmltools::tagList()))

  expect_s3_class(router_server(rt, list(), list(), session), "Observer.event")
})

test_that("router_link", {
  expect_equal(router_link(""), "./#!/")
  expect_equal(router_link("/"), "./#!/")
  expect_equal(router_link("blah"), "./#!/blah")
  expect_equal(router_link("/blah/"), "./#!/blah/")
})


R/shiny.R

#' Router server and UI
#'
#' @description
#' The `router_server()` and `router_ui()` functions are used to render page
#' modules served by a `[router()]` into a Shiny application.
#'
#' The content for the current page will be rendered in place of `router_ui()`.
#' This can appear anywhere inside the UI definition, so that common UI elements
#' like `head()` tags don't need to be repeated in each page module.
#'
#' `router_server()` can be called from anywhere inside the `server()` function
#' of the Shiny application.
#'
#' @param id Router id.
#' @param router A shinypages `router()` object.
#' @param input Shiny server `input` object.
#' @param output Shiny server `output` object.
#' @param session Shiny server `session` object.
#'
#' @return `router_server()` returns a Shiny observer reference class object.
#'   See [shiny::observe()] for details.
#' @return `router_ui()` returns an HTML output element.
#'
#' @name router_shiny
#' @seealso [router()]
#'
#' @examples
#' if (interactive()) {
#'
#' ui_test <- function(id) {
#'   htmltools::div(
#'     htmltools::p("Test!"),
#'     htmltools::p(htmltools::a(href = router_link("/"), "Home"))
#'   )
#' }
#'
#' test_router <- router(router_page("home", ui_test))
#'
#' ui <- shiny::fluidPage(
#'   router_ui()
#' )
#'
#' server <- function(input, output, session) {
#'   router_server(test_router, input, output, session)
#' }
#'
#' shiny::shinyApp(ui, server)
#' }
NULL

#' @rdname router_shiny
#' @export
router_ui <- function(id = "__router__") {
  shiny::uiOutput(id)
}

router_serve_page <- function(router, session) {
  url_path <- get_router_path()

  # Select route
  # - use default if blank
  # - check for matching pages
  # - set page to 404 if nothing else is matched
  if (url_path == "") {
    if (!is.null(router$callbacks$default)) {
      page <- router$routes[[router$callbacks$default(session)]]
    } else {
      page <- router$default
    }
    shiny::updateQueryString(clean_hash(page$path), mode = "replace", session)
  } else {
    page <- do.call(
      switch,
      c(list(url_path),
        router$routes,
        list(router$http$page_404))
    )
  }

  # If not authorised set page to 403
  if (!is.null(page$authorised) && !page$authorised(session)) {
    page <- router$http$page_403
  }

  page
}

#' @rdname router_shiny
#' @export
router_server <- function(router, input, output, session, id = "__router__") {
  stopifnot("`router` is not a shinypages <router> object" = inherits(router, "router"))

  session$userData$router_page <- shiny::reactiveVal()

  shiny::observeEvent(shiny::getUrlHash(session), priority = 1000, {
    page <- router_serve_page(router, session)

    # Update router_page reactive
    session$userData$router_page(page)

    if (!is.null(page$server_module)) {
      page$server_module(paste0(id, router_ns(get_router_page()$path)))
    }

    output[[id]] <- shiny::renderUI(
      page$ui_module(paste0(id, router_ns(get_router_page()$path)))
    )

    if (!is.null(router$callbacks$pageload)) {
      router$callbacks$pageload(page, session)
    }
  })
}

#' Create a page link
#'
#' Take a path to a page and convert it to a link for use in Shiny UI.
#'
#' @param path Path of the destination page.
#'
#' @return A link to the destination page.
#' @export
#'
#' @examples
#' htmltools::a(href = router_link("home"), "Home page")
router_link <- function(path) {
  paste0("./", clean_hash(path))
}

#' Change router page
#'
#' Set the current router page to `path`.
#'
#' @param path Path of the destination page.
#' @param mode Argument passed to [shiny::updateQueryString()].
#' @param session The current shiny session object.
#'
#' @export
#'
#' @examples
#' \dontrun{
#' change_router_page("home")
#' }
change_router_page <- function(path, mode = "push", session = shiny::getDefaultReactiveDomain()) {
  shiny::updateQueryString(clean_hash(path), mode, session)
}

#' Get the current page
#'
#' Get the `router_page()` object for the current page. Can be used to access
#' metadata and other page attributes.
#'
#' @param session The current shiny session object.
#'
#' @return The current [router_page()] object.
#' @export
#'
#' @examples
#' \dontrun{
#' get_router_page()
#' }
get_router_page <- function(session = shiny::getDefaultReactiveDomain()) {
  session$userData$router_page()
}

#' Get the title of the current page
#'
#' Get the title attribute of the current page.
#'
#' @param session The current shiny session object.
#'
#' @return The title of the current [router_page()].
#' @export
#'
#' @examples
#' \dontrun{
#' get_router_page_title()
#' }
get_router_page_title <- function(session = shiny::getDefaultReactiveDomain()) {
  page <- session$userData$router_page
  if (is.null(page))
    NULL
  else
    page()$title
}

#' Get the metadata of the current page
#'
#' Get the list stored in the metadata attribute of the current page.
#'
#' @param session The current shiny session object.
#'
#' @return The metadata list for the current [router_page()].
#' @export
#'
#' @examples
#' \dontrun{
#' get_router_page_metadata()
#' }
get_router_page_metadata <- function(session = shiny::getDefaultReactiveDomain()) {
  page <- session$userData$router_page
  if (is.null(page))
    NULL
  else
    page()$metadata
}

#' Get the path of the current page
#'
#' Extract the current router path from the URL.
#'
#' @param session The current shiny session object.
#'
#' @return The current router path.
#' @export
#'
#' @examples
#' \dontrun{
#' get_router_path()
#' }
get_router_path <- function(session = shiny::getDefaultReactiveDomain()) {
  url_hash <- shiny::getUrlHash(session)
  url_path <- extract_path(clean_hash(url_hash))
  url_path
}


R/router.R

#' Create a router page
#'
#' @description
#' A page attaches a Shiny module to a path, to be included in a [router()]
#' definition.
#'
#' @section Shiny module:
#'
#' The `ui_module` and `server_module` arguments take shiny module UI and server
#' functions respectively. The module functions shouldn't take any arguments
#' apart from `id`.
#'
#' ```r
#' ui_test <- function(id) {
#'   htmltools::p("Test!")
#' }
#'
#' server_test <- function(id) {
#'   shiny::moduleServer(id, function(input, output, session) { })
#' }
#'
#' router_page("test", ui_test, server_test)
#' ```
#'
#' For an introduction to Shiny modules, see the [Modularizing Shiny app
#' code](https://shiny.rstudio.com/articles/modules.html) article.
#'
#' @section Authorising sessions:
#'
#' A page can be configured to check if a session is authorised to view it
#' before serving the page. To check authorisation, pass a function in the
#' `authorised` argument that takes a shiny session as input and returns `TRUE`
#' if a user is authorised to view the page and `FALSE` otherwise. For example:
#'
#' ```r
#' is_admin <- function(session) {
#'   "admin" %in% session$groups
#' }
#' ```
#'
#' The `authorised` function is run by the router before serving the page. If
#' the function returns `FALSE`, the router will serve the 403 Forbidden page.
#'
#' @section Metadata:
#'
#' The page can have two pieces of metadata attached:
#'
#' * A `title`, which can be retrieved inside a Shiny app with
#' [get_router_page_title()].
#'
#' * `metadata`, a list of arbitray data, which can be retrieved inside a Shiny
#' app with [get_router_page_metadata()].
#'
#' @param path Path for this page.
#' @param ui_module Shiny module UI to render.
#' @param server_module Shiny server module to render.
#' @param title Page title. This can be accessed as reactive via
#'   [get_router_page_title()].
#' @param authorised A function that determines whether the current session is
#'   authorised to view this page. The function should take a shiny session as
#'   input, and return `TRUE` if the session is authorised to view this page and
#'   `FALSE` otherwise.
#' @param metadata An arbitrary list of page metadata. This can be accessed as a
#'   reactive via [get_router_page_metadata()] or used in router callbacks.
#'
#' @return A `router_page` object.
#' @seealso [router()]
#' @export
#' @examples
#' if (interactive()) {
#'
#' ui_test <- function(id) {
#'   htmltools::p("Test!")
#' }
#'
#' server_test <- function(id) {
#'   shiny::moduleServer(id, function(input, output, session) { })
#' }
#'
#' home_page <- router_page("home", ui_test)
#' server_page <- router_page("server", ui_test, server_test)
#'
#' is_admin <- function(session) {
#'   "admin" %in% session$groups
#' }
#'
#' admin_page <- router_page("admin", ui_test, authorised = is_admin)
#' }
router_page <- function(path,
                        ui_module,
                        server_module = NULL,
                        title = NULL,
                        authorised = NULL,
                        metadata = list()) {
  stopifnot("`path` is not a character value" = is.character(path))
  stopifnot("`ui_module` is not a function" = is.function(ui_module))
  stopifnot("`server_module` is not a function" = is.function(server_module) | is.null(server_module))
  stopifnot("`title` is not a character value" = is.character(title) | is.null(title))
  stopifnot("`authorised` is not a function" = is.function(authorised) | is.null(authorised))
  stopifnot("`metadata` is not a list" = is.list(metadata))

  structure(
    list(
      path = path,
      ui_module = ui_module,
      server_module = server_module,
      title = title,
      authorised = authorised,
      metadata = metadata
    ),
    class = "router_page"
  )
}

#' @rdname router_page
#' @param x A `router_page` object.
#' @param ... Additional arguments for specific methods.
#' @param short Should a one line summary be printed?
#' @export
print.router_page <- function(x, short = FALSE, ...) {
  if (has_cli()) {
    cli::cli_text("{.cls {class(x)}} {.emph #!/{x$path}} {if (!is.null(x$title)) paste0(' - ', x$title)}")

    if (!short) {
      tick_or_cross(!is.null(x$server_module), "Server module")
      tick_or_cross(!is.null(x$authorised), "Authorisation function")
      cli::cli_text("Has {cli::no(length(x$metadata))} metadata item{?s}")
    }
  } else {
    cat("<", class(x), "> #!/", x$path, if (!is.null(x$title)) paste0(' - ', x$title), "\n", sep = "")
  }

  invisible(x)
}

#' Create a router
#'
#' @description
#'
#' A [router()] defines a multi-page Shiny application using a set of
#' [router_page()]s. See [router_server()] and [router_ui()] to create a
#' multi-page application from the returned `router` object.
#'
#' @section Callbacks:
#'
#' The callback function arguments allow user customisation of the page serving
#' process.
#'
#' * `callback_pageload` is called after a new page is loaded. It should be a
#' function with the signature `function(page, session)`, taking a `router_page`
#' and a shiny session as input. Thh return value of the function is ignored.
#'
#'   Useful to perform operations based on page metadata.
#'
#'   ```r
#'   callback_pageload = function(page, session) {
#'     shinyjs::runjs(paste0("document.title = \"", page$title, "\";"))
#'   }
#'   ```
#'
#' * `callback_default()` is called to determine the path to serve when the
#' default page is requested. It should be a function with the signature
#' `function(session)`, taking a shiny session as input and returning the path
#' of the page to load.
#'
#'   Useful for setting per-user or per-group home pages.
#'
#'   ```r
#'   callback_default = function(session) {
#'     if ("dev" %in% session$groups) {
#'       "devlanding"
#'     } else {
#'       "home"
#'     }
#'   )
#'   ```
#'
#' @section HTTP response pages:
#'
#' `router_page`s are served in place of common HTTP response codes. These can
#' be customised:
#'
#' * `page_403` is served when the `authorised` function of a page returns
#' `FALSE`.
#'
#' * `page_404` is served when an unknown path is requested.
#'
#' @param ... A set of [router_page()] objects. The first page is considered the
#'   default, and will be used if no path is provided.
#' @param callback_pageload A function that is called immediately after page
#'   content is rendered. The function should take a `router_page` object and
#'   shiny session as input.
#' @param callback_default A function that determines the default page. The
#'   function should take a shiny session as input and return the path of the
#'   default page. If not provided, the first page is used as the default.
#' @param page_403 A `router_page` to be used when a user attempts to access an
#'   unauthorised page.
#' @param page_404 A `router_page` to be used when a user attempts to access a
#'   page that doesn't exist.
#'
#' @return A `router` object.
#' @export
#' @seealso [router_page()] to create pages, [add_router_callback()] to add
#'   callback functions to an existing router, and [router_server()] and
#'   [router_ui()] to add a router to a Shiny application.
#' @examples
#' if (interactive()) {
#'
#' ui_test <- function(id) {
#'   htmltools::p("Test!")
#' }
#'
#' home_page <- router_page("home", ui_test)
#' other_page - router_page("other", ui_test)
#'
#' test_router <- router(home_page, other_page)
#' }
router <- function(...,
                   callback_pageload = NULL,
                   callback_default = NULL,
                   page_403 = router_page("403", ui_403),
                   page_404 = router_page("404", ui_404)) {
  routes <- list(...)
  stopifnot("Only <router_page> objects should be provided in `...`" = all(vapply(routes, function(x) inherits(x, "router_page"), logical(1))))
  stopifnot("Some paths are duplicated" = !any(duplicated(vapply(routes, function(x) x$path, character(1)))))
  stopifnot("At least 1 page must be provided in `...`" = length(routes) > 0)

  x <- structure(
    list(
      routes = list(...)
    ),
    class = "router"
  )

  names(x$routes) <- sapply(x$routes, function(x) { x$path })
  x$default <- x$routes[[1]]
  x$current <- new.env(parent = emptyenv())
  x$callbacks <- list(pageload = callback_pageload, default = callback_default)
  x$http <- list(page_403 = page_403, page_404 = page_404)

  x
}

#' @rdname router
#' @param x A `router` object.
#' @param short Should a one line summary be printed?
#' @export
print.router <- function(x, short = FALSE, ...) {
  paths <- sapply(x$routes, function(x) x$path)
  if (has_cli()) {
    cli::cli_text("{.cls {class(x)}} with {cli::no(length(x$routes))} page{?s/:/s:} {.emph {paths}}")
    cli::cli_text("{.strong Callbacks:}")
    tick_or_cross(!is.null(x$callbacks$pageload), "pageload")
    tick_or_cross(!is.null(x$callbacks$default), "default")

    if (!short) {
      cli::cli_text("{.strong Pages:}")
      lapply(x$routes, print, short = TRUE)
    }
  } else {
    cat("<", class(x), "> with ", length(x$routes), " page(s): ", paste0(paths, collapse = ", "), "\n", sep = "")
    if (!short) {
      cat("Pages:\n")
      lapply(x$routes, print, short = TRUE)
    }
  }

  invisible(x)
}

#' Set a router callback function
#'
#' @param router A router object.
#' @param callback The name of the callback to set. Currently `"pageload"` and
#'   `"default"` callbacks are supported.
#' @param func The callback function. The function should take a `router_page`
#'   object and shiny session as input.
#'
#' @return Invisibly returns the input `router` object.
#' @export
#' @seealso [router()]
#' @examples
#' \dontrun{
#' add_router_callback(
#'   router,
#'   "pageload",
#'   function(page, session) {
#'     shinyjs::runjs(paste0("document.title = \"", page$title, "\";"))
#'   }
#' )
#'
#' add_router_callback(
#'   router,
#'   "default",
#'   function(session) {
#'     if ("dev" %in% session$groups) {
#'       "devlanding"
#'     } else {
#'       "home"
#'     }
#'   }
#' )
#' }
add_router_callback <- function(router, callback = c("pageload", "default"), func) {
  stopifnot("`router` is not a shinypages <router> object" = inherits(router, "router"))
  stopifnot("`func` is not a function" = is.function(func))
  callback <- match.arg(callback)

  router$callbacks[[callback]] <- func
  invisible(router)
}
